python/src/uagents/registration.py

```diff
@@ -40,3 +61,162 @@ def _build_digest(self) -> bytes:
+                        f"Adding testnet funds to {self._wallet.address()}"
+                    )
+                else:
+                    self._logger.info(
+                        f"Send funds to wallet address: {self._wallet.address()}"
+                    )
+                raise InsufficientFundsError()
+
+            self._logger.info("Registering on almanac contract...")
+
+            signature = self._sign_registration(agent_address)
+            await self._almanac_contract.register(
+                self._ledger,
+                self._wallet,
+                agent_address,
+                protocols,
+                endpoints,
+                signature,
+            )
+            self._logger.info("Registering on almanac contract...complete")
+        else:
+            self._logger.info("Almanac registration is up to date!")
+
+    def _get_balance(self) -> int:
+        return self._ledger.query_bank_balance(Address(self._wallet.address()))
+
+    def _sign_registration(self, agent_address: str) -> str:
+        """
+        Sign the registration data for Almanac contract.
+
+        Returns:
+            str: The signature of the registration data.
+
+        Raises:
+            AssertionError: If the Almanac contract address is None.
+
+        """
+        assert self._almanac_contract.address is not None
+        return self._identity.sign_registration(
+            str(self._almanac_contract.address),
+            self._almanac_contract.get_sequence(agent_address),
+        )
+
+
+class DefaultRegistrationPolicy(AgentRegistrationPolicy):
+    def __init__(
+        self,
+        identity: Identity,
+        ledger: LedgerClient,
+        wallet: LocalWallet,
+        almanac_contract: AlmanacContract,
+        testnet: bool,
+        *,
+        logger: Optional[logging.Logger] = None,
+        almanac_api: Optional[str] = None,
+    ):
+        self._logger = logger or logging.getLogger(__name__)
+        self._api_policy = AlmanacApiRegistrationPolicy(
+            identity, almanac_api=almanac_api, logger=logger
+        )
+        self._ledger_policy = LedgerBasedRegistrationPolicy(
+            identity, ledger, wallet, almanac_contract, testnet, logger=logger
+        )
+
+    async def register(
+        self,
+        agent_address: str,
+        protocols: List[str],
+        endpoints: List[AgentEndpoint],
+    ):
+        # prefer the API registration policy as it is faster
+        try:
+            await self._api_policy.register(agent_address, protocols, endpoints)
+        except Exception as e:
+            self._logger.error(f"Failed to register on Almanac API: {e}")
+
+        # schedule the ledger registration
+        try:
+            await self._ledger_policy.register(agent_address, protocols, endpoints)
+        except InsufficientFundsError:
+            self._logger.warning(
+                "Failed to register on Almanac contract due to insufficient funds"
+            )
```

Comment:
We also need to reraise here in order to schedule the retry after a few seconds instead of the default registration interval.